src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs

```diff
@@ -63,14 +64,47 @@ public override void AdvanceTo(SequencePosition consumed)
 
         public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
         {
-            var dataLength = _readResult.Buffer.Slice(_readResult.Buffer.Start, consumed).Length;
+            // This code path is fairly hard to understand so let's break it down with an example
+            // ReadAsync returns a ReadResult of length 50.
+            // Advance(25, 40). The examined length would be 40 and consumed length would be 25.
+            // _totalExaminedInReadResult starts at 0. newlyExamined is 40.
+            // OnDataRead is called with length 40.
+            // _totalExaminedInReadResult is now 40 - 25 = 15.
```

Comment:
So _totalExaminedInReadResult is the data already examined in the next read result, not this one. That's why you can subtract out consumed.

Reply:
I'll rename the variable.